File: SteffensenMethod.py
#!/usr/bin/env python

# finds an approximation of a root using Newton's Method
# default iterations is 15

from sympy import *
import sys

if len(sys.argv) != 3 :
    print("steffensen [function] [seed]")
    exit()

# function:string, seed:number
def steffensenMethod (function, seed, precision) :

    # error check
    if precision == 0 :
        exit()
    
    # setup
    h = precision
    x = seed
    f = lambda x : eval(function)
    g = lambda x : (f(x+h)-f(x))/h
    iterations = 50

    while (iterations > 0) :
        x -= f(x)/float(g(x))
        iterations -= 1
    return x
    
function = sys.argv[1]
seed = eval(sys.argv[2])

# solution close to
print ((steffensenMethod(function,seed,0.1)))

File: steffensen_method/steffensenMethod.c
//
#include <stdio.h>
#include <stdlib.h>

double function(double x)
{
    return x * x;
}

double steffensenMethod(double(*f)(double), 
    double seed, double precision, int iterations)
{
    if(precision == 0)
    {
        exit(-1);
    }

    double h = precision;
    double x = seed;
    // legal gnu c
    double g(double x)
    {
        return (f(x + h) - f(x)) / h;
    }

    int i = iterations;

    while(i > 0)
    {
        x -= f(x) / ((float)(g(x)));
        i -= 1;
    }
    return x;

}


int main()
{
    printf("%f\n", 
        steffensenMethod(&function, 3, 0.1, 50));
    return 0;
}
